### PR TITLE
SSE: Fix math expression to support NoData results

### LINKED ISSUE #61304: Alert rule panic; runtime error: invalid memory address or nil pointer dereference
**What happened**:
- Grafana quits unexpectedly from time to time, logging alert rule panic (stacktrace below)

**What you expected to happen**:
- For it not to crash :wink: 

**How to reproduce it (as minimally and precisely as possible)**:
I have the alert rule that I pasted below, if I click the Refresh button, it shows an error popup in the top right, I think this might be the cause. The screenshot below is of the rule with the uid from the stack trace.

**Anything else we need to know?**:
- Might be related to #58703
- [Stack trace](https://github.com/grafana/grafana/files/10393720/stack.txt)

**Environment**:
- Grafana version: v9.3.2 \(21c1d14e91\)
- Data source type & version: AlertManager, Athena, cloudwatch, ElasticSearch, OpenSearch, Prometheus
- OS Grafana is installed on: Linux (AWS ECS)
- User OS & Browser: EndeavourOS; Firefox 108.0.2
- Grafana plugins: 
  - Amazon Athena
  - Direct Input
  - Grafana OnCall
  - OpenSearch
- Others:

image

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
Currently, when NoData result is passed to any unary operator it returns a slice `[]Value{nil}` which causes panic in subsequent operations because `switch` statements that analyze the type of the Value do not expect nil. 

This PR updates all unary operations to return NoData if the input argument has type NoData.

A union operation produces an empty `Union` when either of the results is NoData. This makes the math expression result in empty values slice. This PR updates the union operation to return a non empty union where at least one of values is NoData and any binary operation that uses the union to return NoData when either of the arguments is NoData.

**Why do we need this feature?**

This will make NoData be correctly propagated through the pipeline

**Who is this feature for?**

Grafana Alerting users are the primary beneficiary of this change

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/61304

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/expr/mathexp/exp_nan_null_val_test.go [hunk 1]
```diff
@@ -422,3 +424,89 @@ func TestNullValues(t *testing.T) {
...
+					require.Len(t, res.Values, 1)
+					require.Equal(t, NewNoData(), res.Values[0])
+				})
+
+				t.Run("$A=series, $B=nodata", func(t *testing.T) {
+					res, err := e.Execute("", makeVars(NewNoData(), series))
+					require.NoError(t, err)
+					require.Len(t, res.Values, 1)
+					require.Equal(t, NewNoData(), res.Values[0])
+				})
```
Comment (kylebrandt) [inline pkg/expr/mathexp/exp_nan_null_val_test.go:508]: What happens with `$A=nodata,$B=nodata`, panic?

Comment (yuri-tceretian) [reply pkg/expr/mathexp/exp_nan_null_val_test.go:508]: nodata, there is a test above (line 489)

**Review by kylebrandt — APPROVED** [review]
Looks good to me, thank you for either fixing, or getting us much closer to to having No data fixed! Would be good to add a bit about no data in https://github.com/grafana/grafana/blob/main/docs/sources/panels-visualizations/query-transform-data/expression-queries/index.md#operators

**Comment by Roberto6969** [comment]
Hello @yuri-tceretian 
despite the upgrade to version=9.4.0-97658pre, I'm still getting an error

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x27a27f8]

Should I change/repair anything else?

Thank You!

**Comment by yuri-tceretian** [comment]
Hi Roberto, what is the stack trace? I am pretty sure you encounter something different problem than this PR fixes.

**Comment by Roberto6969** [comment]
@yuri-tceretian thank you for your feedback
Grafana crashed after update on 17.1., late evening. I believe it will be easier if I attach Log  
[_grafana_logs.txt](https://github.com/grafana/grafana/files/10468612/_grafana_logs.txt)

**Comment by yuri-tceretian** [comment]
Can you tell from what version you migrated?

**Comment by Roberto6969** [comment]
From same version - 9.4.0, but lower prelease number

**Comment by Roberto6969** [comment]
Hi @yuri-tceretian I've found solution. If I downgrade version to 9.3.2 and restore grafana.db, Grafana works fine. But whenever I upgrade to latest version, it crashes with errors which I've mentioned before. I'm afraid this problem is based on alerting. Also, I have following errors:

logger=ngalert.state.historian rule_uid=ZW61rXzRz org_id=2 t=2023-01-23T06:24:04.67531922Z level=error msg="Error getting dashboard for alert annotation" dashboardUID=gicgdhh error="Dashboard not found"

logger=ngalert.state.manager t=2023-01-23T06:24:06.190140254Z level=warn msg="Rule declares one or many reserved labels. Those rules labels will be ignored" labels="alertname=Moj Telegram" 

How can I resolve this problems? 
Thank You

**Comment by Roberto6969** [comment]
@yuri-tceretian Indeed - problem is fixed. I've updated Grafana to 9.4.0 successfully! Now is OK. Thank You a lot!

Is this error in log because of my settings or because of bug in Grafana:

logger=ngalert.sender.router rule_uid=ZW61rXzRz org_id=2 t=2023-01-25T14:49:07.978594529Z level=info msg="Sending alerts to local notifier" count=1
logger=ngalert.state.historian rule_uid=ZW61rXzRz org_id=2 t=2023-01-25T14:49:07.990627144Z level=error msg="Error getting dashboard for alert annotation" dashboardUID=gicgdhh error="Dashboard not found"

How can I find where is this rule_uid=ZW61rXzRz?

**Pull request merged**